crawlers/mooncrawl/eth_scan.py

```diff
@@ -0,0 +1,184 @@
+from moonstreamdb.models import EthereumAddress, EthereumLabel
+import requests
+
+
+ETH_SCAN_TOKEN = os.environ.get("MOONSTREAM_ETHERSCAN_TOKEN")
+if ETH_SCAN_TOKEN is None:
+    raise ValueError("MOONSTREAM_ETHERSCAN_TOKEN environment variable must be set")
+
+BASE_API_URL = "https://api.etherscan.io/api?module=contract&action=getsourcecode"
+
+ETHERSCAN_SMARTCONTRACTS_LABEL_NAME = "etherscan_smartcontract"
+
+bucket = os.environ.get("AWS_S3_SMARTCONTRACT_BUCKET")
+if bucket is None:
+    raise ValueError("AWS_S3_SMARTCONTRACT_BUCKET must be set")
+
+
+@dataclass
+class VerifiedSmartContract:
+    name: str
+    address: str
+    tx_hash: str
+
+
+def push_to_bucket(contract_data: Dict[str, Any], contract_file: str):
+    result_bytes = json.dumps(contract_data).encode("utf-8")
+    result_key = contract_file
+
+    s3 = boto3.client("s3")
+    s3.put_object(
+        Body=result_bytes,
+        Bucket=bucket,
+        Key=result_key,
+        ContentType="application/json",
+        Metadata={"source": "etherscan", "crawler_version": ""},
+    )
+
+
+def get_address_id(db_session: Session, contract_address: str) -> int:
+    """
+    Searches for given address in EthereumAddress table,
+    If doesn't find one, creates new.
+    Returns id of address
+    """
+    query = db_session.query(EthereumAddress.id).filter(
+        EthereumAddress.address == contract_address
+    )
+    id = query.one_or_none()
+    if id is not None:
+        return id[0]
+
+    latest_address_id = (
+        db_session.query(EthereumAddress.id).order_by(text("id desc")).limit(1).one()
+    )[0]
+
+    id = latest_address_id + 1
+    smart_contract = EthereumAddress(
+        id=id,
+        address=contract_address,
+    )
+    db_session.add(smart_contract)
+    db_session.commit()
+    return id
+
+
+def crawl_step(db_session: Session, contract: VerifiedSmartContract, crawl_url: str):
+    attempt = 0
+    current_interval = 2
+    success = False
+
+    response: Optional[requests.Response] = None
+    while (not success) and attempt < 3:
+        attempt += 1
+        try:
+            response = requests.get(crawl_url)
+            response.raise_for_status()
+            success = True
+        except:
+            current_interval *= 2
+            time.sleep(current_interval)
+
+    if response is None:
+        print(f"Could not process URL: {crawl_url}", file=sys.stderr)
+        return None
+    page = response.json()
+    result = page["result"][0]
+    contract_info = {
+        "data": result,
+        "crawl_version": MOONCRAWL_VERSION,
+        "crawled_at": f"{datetime.now()}",
+    }
+    object_key = f"/v1/etherscan/{contract.address}.json"
+    push_to_bucket(contract_info, object_key)
+
+    eth_address_id = get_address_id(db_session, contract.address)
+
+    eth_label = EthereumLabel(
+        label=ETHERSCAN_SMARTCONTRACTS_LABEL_NAME,
+        address_id=eth_address_id,
+        label_data={
+            "key": object_key,
```

Comment:
Instead of key, can you put the full `s3_uri` here?

`"s3_uri": f"s3://{bucket}/{object_key}`.

This way we are okay even if we change the bucket in the future.

Reply:
Later, we may even want to switch from AWS S3 to Google Cloud Storage. I don't mind if we use the `s3_uri` key for both, but if you want to be really future proof, call it something like `"object_uri"`.